servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +173,101 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
+
+        RequestTimeoutSubscriber(Subscriber<? super StreamingHttpResponse> delegate, Completable requestComplete,
+                                        Duration timeout, Executor timeoutExecutor) {
+            this.delegate = delegate;
+            this.requestComplete = requestComplete;
+            this.timeout = timeout;
+            this.timeoutExecutor = timeoutExecutor;
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            delegate.onSubscribe(() -> {
+                // We don't need to condition cancellation here on the `once()` call because the `DelayedCancellable`
+                // will already enforce idempotency of the cancel call, and it's fine if we're racing cancels with the
+                // results since cleanup is gated on the `once()` call.
+                once();
+                timeoutCancellable.cancel();
+                requestCancellable.cancel();
+            });
+            requestCancellable.delayedCancellable(cancellable);
+            timeoutCancellable.delayedCancellable(requestComplete.concat(Completable.never()
+                    .timeout(timeout, timeoutExecutor)).beforeOnError(this::handleInterruptions).subscribe());
+        }
+
+        private void handleInterruptions(Throwable t) {
+            if (once()) {
+                requestCancellable.cancel();
+                Throwable result = t;
+                // We can get a SocketTimeoutException waiting for a 100 Continue response.
+                if (t instanceof TimeoutException) {
+                result = newStacklessSocketTimeoutException(
+                        "Read timed out after " + timeout.toMillis() +
+                                "ms waiting for response meta-data")
+                        .initCause(t);
+                }
+                delegate.onError(result);
+            }
+        }
+
+        @Override
+        public void onSuccess(@Nullable StreamingHttpResponse result) {
+            if (once()) {
+                timeoutCancellable.cancel();
+                if (result != null) {
+                    result = result.transformMessageBody(p -> p.timeout(timeout, timeoutExecutor)
+                            .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
+                                    "Read timed out after " + timeout.toMillis() +
+                                            "ms waiting for the next response payload body chunk")
+                                    .initCause(t)));
+                }
+                delegate.onSuccess(result);
+            } else {
+                if (result != null) {
+                    toSource(result.messageBody()).subscribe(CancelImmediatelySubscriber.INSTANCE);
+                }
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            if (once()) {
+                timeoutCancellable.cancel();
+                delegate.onError(t);
```

Comment:
We should protect from unexpected exceptions from `cancel()` like we do inside `AmbSingles.State`